pkg/lib/authn/authenticator/password/provider.go

```diff
@@ -95,13 +96,14 @@ type UpdatePasswordOptions struct {
 // Otherwise original authenticator will be returned
 func (p *Provider) UpdatePassword(a *authenticator.Password, options *UpdatePasswordOptions) (bool, *authenticator.Password, error) {
 	password := options.PlainPassword
-	err := p.PasswordChecker.ValidateNewPassword(a.UserID, password)
-	if err != nil {
-		return false, nil, err
-	}
 
 	newAuthen := a
-	if pwd.Compare([]byte(password), a.PasswordHash) != nil {
+	if options.SetPassword && pwd.Compare([]byte(password), a.PasswordHash) != nil {
+		err := p.PasswordChecker.ValidateNewPassword(a.UserID, password)
+		if err != nil {
+			return false, nil, err
+		}
+
```

Comment:
The logic changed. It should be kept unchanged. We always validate the password first.

```golang
if options.SetPassword {
  err := p.PasswordChecker.ValidateNewPassword(a.UserID, password)
  if err != nil {
    return false, nil, err
  }
  if pwd.Compare([]byte(password), a.PasswordHash) != nil {
    newAuthen = p.populatePasswordHash(a, password)
  }
}
```